File: probs/binary-tree/105.construct-binary-tree-from-preorder-and-inorder-traversal.py
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        # preorder = node, lefts, rights
        # inorder = left, node, rights

        if len(preorder) == 0 or len(inorder) == 0:
            return
        N = len(preorder)
        node = preorder[0]
        if node not in inorder:
            return
        L = inorder.index(node)

        return TreeNode(
            node,
            self.buildTree(preorder[1:], inorder[: L + 1]),
            self.buildTree(preorder[L + 1 :], inorder[L + 1 :]),
        )

File: probs/binary-tree/105.construct-binary-tree-from-preorder-and-inorder-traversal.java
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder.length == 0 || inorder.length == 0) return null;
        // preorder = node -> lefts -> rights 
        // inorder = lefts -> node -> rights
        int node = preorder[0];
        int N = preorder.length;
        
        int L = IntStream.range(0, inorder.length)
            .filter(i -> inorder[i] == node).findFirst().orElse(-1);

        if (L < 0) return null;
        
        // for lefts, reconstruct 
        TreeNode left = buildTree(Arrays.copyOfRange(preorder, 1, L+1), Arrays.copyOfRange(inorder, 0, L));
        TreeNode right = buildTree(Arrays.copyOfRange(preorder, L+1, N), Arrays.copyOfRange(inorder, L+1, N));
        return new TreeNode(node, left, right);
    }
}
